refactor(levm): rewrite of state EF tests runner first iteration

sofiazcoaga opened the pull request:
**Motivation**

Related issue: #3496. 

The idea is to incrementally develop a new EF Test runner (for state tests) that can eventually replace the current one. The main goal of the new runner is to be easy to understand and as straightforward as possible, also making it possible to easily add any new requirement. 

**How to run** 
A target in the makefile was included. You can, then, from `ethrex/cmd/ef_tests/state/` run `make run-new-runner`. If no specific path is passed, it will parse anything in the `./vectors` folder. Otherwise you can do, for example: 
 `make run-new-runner TESTS_PATH=./vectors/GeneralStateTests/Cancun` to specify a path. 

This command assumes you have the `vectors` directory downloaded, if not run `make download-evm-ef-tests` previously. 

**Considerations**

The main changes are: 
- The new `Test` and `TestCase` structures in types. 
- The runner and parser simplified flows. 

Files that should not be reviewed as they are full or partial copies of the original files: 
- `runner_v2/deserialize.rs`
- `runner_v2/utils.rs`

This iteration excludes report-related code, option flags and other possible test case errors to be considered that will be included later. Checks are performed only on exceptions and root hash.

cdiielsi reviewed (approved):
Really nice work! I like how you distributed the modules, having runner separated from result_check makes things feel less crowded and easier to navigate. I made a couple of comments, one you can consider moving on (it's about writing reports) the rest are rather subjective and totally skippable.

sofiazcoaga commented on cmd/ef_tests/state/runner_v2/result_check.rs:
```diff
@@ -0,0 +1,185 @@
+    Ok(())
+}
+
+fn exception_is_expected(
```
Certainly a more declarative name could be better. Because this is a part of the former runner I still haven't looked at it in depth, but I provisionally changed the name to make it a bit clearer [here]().

sofiazcoaga commented on cmd/ef_tests/state/runner_v2/error.rs:
```diff
@@ -0,0 +1,24 @@
+    FailedToDeserializeField(String),
+    FailedToCreateReportFile(String),
+    FailedToGetIndexValue(String),
+}
```
I think this might be a good suggestion, for now I think it could be considered as out of scope as the new runner is still in a quite rustic state, but I will keep in mind for later iterations to improve the Errors.

JereSalo commented:
Suggestion, leave in PR description the necessary steps for running the tests

sofiazcoaga commented:
@JereSalo your suggestions where added here: 
- The parsing error was fixed by making state field optional [here](https://github.com/lambdaclass/ethrex/pull/3642/commits/49715aeb442a8d6e2f5fbe08df3ac27efa658642)
- Include the possibility to specify a path to the tests [here](https://github.com/lambdaclass/ethrex/pull/3642/commits/f9df85c1dc830a176acd98036f76300799e4c261)
- Makefile command to run the tests [here](https://github.com/lambdaclass/ethrex/pull/3642/commits/6d790a6bd72226cc2b39eaf7b67c99c4c4bf336f)
- How to run the tests left in the PR description 

The comment about repeated runs I consider better to leave for another PR. What do you think?

JereSalo commented on cmd/ef_tests/state/runner_v2/parser.rs:
```diff
@@ -0,0 +1,65 @@
+    "create2collisionStorageParis.json", // Skip because it fails on REVM
+    "InitCollisionParis.json",         // Skip because it fails on REVM
+    "InitCollision.json",              // Skip because it fails on REVM
+    "contract_create.json",
```
Add a comment here

sofiazcoaga replied:
Added!

JereSalo commented on cmd/ef_tests/state/runner_v2/runner.rs:
```diff
@@ -0,0 +1,114 @@
+        let (mut db, initial_block_hash, storage) = load_initial_state(test).await;
+        let env = get_vm_env_for_test(test.env, test_case)?;
+        let tx = get_tx_from_test_case(test_case)?;
+        let tracer = LevmCallTracer::disabled();
+        let vm_type = ethrex_levm::vm::VMType::L1;
+        let mut vm = VM::new(env.clone(), &mut db, &tx, tracer, vm_type);
```
```suggestion
        let tracer = LevmCallTracer::disabled(); // Maybe in the future we can use this
        let mut vm = VM::new(env, &mut db, &tx, tracer, VMType::L1);
```

JereSalo commented on cmd/ef_tests/state/runner_v2/runner.rs:
```diff
@@ -0,0 +1,114 @@
+        let execution_report = vm.execute();
+        let res = check_test_case_results(
+            &mut vm,
+            initial_block_hash,
+            storage,
```
Doesn't the db have this info? And `vm` has a `db`

JereSalo replied:
Update on this: we talked about it and the db should have that info but maybe accessing it is not very straightforward because the only way is through `get_block_hash(block_number)` and we just need to know the block number. It's very likely that we don't have that information within the other parameters so it's good to keep it as it is.

sofiazcoaga replied:
I couldn't find a straightforward way to access this from the DB, but I can put more effort into it in another PR and see if it can be improved, what do you think?

JereSalo replied:
It's not important. It's perfectly fine this way

JereSalo commented on cmd/ef_tests/state/runner_v2/runner.rs:
```diff
@@ -0,0 +1,114 @@
+        .iter()
+        .map(|list_item| (list_item.address, list_item.storage_keys.clone()))
+        .collect();
+    let tx = match &test_case.authorization_list {
```
Can you add a comment close to this that explains briefly why we only use 2 kinds of transactions?
What happens with Type 3 transactions? (4844) Aren't there any tests for them?
(I know it's a thing that was present in the other runner but it shoud be in a comment)

JereSalo commented on cmd/ef_tests/state/runner_v2/utils.rs:
```diff
@@ -0,0 +1,51 @@
+pub fn effective_gas_price(test_env: &Env, test_case: &TestCase) -> Result<U256, RunnerError> {
+    match test_case.gas_price {
+        None => {
+            let current_base_fee = test_env
+                .current_base_fee
+                .ok_or(RunnerError::CurrentBaseFeeMissing)?;
+            let priority_fee = test_case
+                .max_priority_fee_per_gas
+                .ok_or(RunnerError::MaxPriorityFeePerGasMissing)?;
+            let max_fee_per_gas = test_case
+                .max_fee_per_gas
+                .ok_or(RunnerError::MaxFeePerGasMissing)?;
```
Can this happen with the `.json` files that we have? If not I'd just unwrap here

JereSalo commented on cmd/ef_tests/state/runner_v2/result_check.rs:
```diff
@@ -0,0 +1,180 @@
+    Ok(())
+}
+
+fn exception_matches_expected(
```
If you find a way of making this function shorter and more readable (while conserving the same behavior) it would be good. Otherwise it's perfectly fine

JereSalo commented on cmd/ef_tests/state/Makefile:
```diff
@@ -159,3 +159,8 @@ samply-benchmarks: ## ⚡️ Run benchmarks and create samply profiling file
 	$(call run_samply,ERC20Mint,REPETITIONS_SLOW,BENCH_MINT_ITERATIONS)
+
+## New EF state tests runner
+TESTS_PATH := ./vectors/GeneralStateTests
```
```suggestion
TESTS_PATH := ./vectors
```
Doesn't it work with all of them?

sofiazcoaga replied:
Yes it does! [Fixed here](https://github.com/lambdaclass/ethrex/pull/3642/commits/ab66cbce6cdeeadff888053d2a18a86ea66e6068)

JereSalo reviewed (changes requested):
Some additional comments:
- Leave rustdocs comments in top of functions/methods. Even the ones that are kinda self-explanatory. Usually one line is enough for some of the short functions.
- Add `runner_report.txt` to `.gitignore`
- Use unwrap and delete a lot of the errors, keep only the errors that are specific to the VM if possible. If something shouldn't happen (like failing to write a file) just unwrap. Abuse this as code clarity and size matters, we want to make it easier and faster to read.

JereSalo commented:
> @JereSalo your suggestions where added here:
> 
> * The parsing error was fixed by making state field optional [here](https://github.com/lambdaclass/ethrex/pull/3642/commits/49715aeb442a8d6e2f5fbe08df3ac27efa658642)
> * Include the possibility to specify a path to the tests [here](https://github.com/lambdaclass/ethrex/pull/3642/commits/f9df85c1dc830a176acd98036f76300799e4c261)
> * Makefile command to run the tests [here](https://github.com/lambdaclass/ethrex/pull/3642/commits/6d790a6bd72226cc2b39eaf7b67c99c4c4bf336f)
> * How to run the tests left in the PR description
> 
> The comment about repeated runs I consider better to leave for another PR. What do you think?

That's fine, I just thought this was important because it's a thing that we should consider for the structure of the tests. But I agree it should be tackle in another PR, it is a priority though.

sofiazcoaga commented on cmd/ef_tests/state/runner_v2/runner.rs:
```diff
@@ -0,0 +1,114 @@
+};
+
+pub async fn run_tests(tests: Vec<Test>) -> Result<(), RunnerError> {
+    // Remove previous report if it exists.
+    let _ = fs::remove_file("./runner_v2/runner_report.txt").await;
```
The thing is, every test report is written as soon as it finishes executing, so I have to open the file in append mode. To do that, I need to make sure that I'm appending to a file that belongs only to this run and no other.

sofiazcoaga commented on cmd/ef_tests/state/runner_v2/run.rs:
```diff
@@ -0,0 +1,15 @@
+    let tests = parse_dir(path.into())?;
+    println!("Finalized parsing. Executing tests...");
+    run_tests(tests).await?;
+    println!("Tests finalized running. Find the report at: './runner_v2/runner_report.txt'");
```
[Fixed](https://github.com/lambdaclass/ethrex/pull/3642/commits/64fc74636ab466f4f31c21ff2c8653b48ed55aee)! :)

sofiazcoaga commented on cmd/ef_tests/state/runner_v2/run.rs:
```diff
@@ -0,0 +1,15 @@
+    let path = &args[1];
+    println!("Parsing test files...");
+    let tests = parse_dir(path.into())?;
+    println!("Finalized parsing. Executing tests...");
```
[Fixed](https://github.com/lambdaclass/ethrex/pull/3642/commits/64fc74636ab466f4f31c21ff2c8653b48ed55aee) :)

sofiazcoaga commented:
> Some additional comments:
> 
> * Leave rustdocs comments in top of functions/methods. Even the ones that are kinda self-explanatory. Usually one line is enough for some of the short functions.
> * Add `runner_report.txt` to `.gitignore`
> * Use unwrap and delete a lot of the errors, keep only the errors that are specific to the VM if possible. If something shouldn't happen (like failing to write a file) just unwrap. Abuse this as code clarity and size matters, we want to make it easier and faster to read.

I removed a lot of custom errors for situations external to VM problems in [this commit](https://github.com/lambdaclass/ethrex/pull/3642/commits/5b43fb72b65a566f7199910c83f88419a0242ec6), replacing with `unwrap()` as you suggested. This made the `error.rs` file smaller as well. 

I added the report file to `.gitignore` [here](https://github.com/lambdaclass/ethrex/pull/3642/commits/ca2faba6808dcb2f55c8e4acc2f7b54831b01521)

JereSalo reviewed (approved):
good job, I think the comment on the type of the transaction is wrong though but the pr is okay
